src/overlay/TxPullMode.cpp

```diff
@@ -0,0 +1,195 @@
+// Copyright 2022 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "overlay/TxPullMode.h"
+#include "crypto/Hex.h"
+#include "ledger/LedgerManager.h"
+#include "overlay/Peer.h"
+#include "util/Logging.h"
+
+namespace stellar
+{
+
+constexpr uint32 const ADVERT_CACHE_SIZE = 50000;
+
+TxPullMode::TxPullMode(Application& app, std::weak_ptr<Peer> peer)
+    : mApp(app)
+    , mAdvertHistory(ADVERT_CACHE_SIZE)
+    , mAdvertTimer(app)
+    , mWeakPeer(peer)
+{
+}
+
+void
+TxPullMode::flushAdvert()
+{
+    if (mOutgoingTxHashes.size() > 0)
+    {
+        StellarMessage adv;
+        adv.type(FLOOD_ADVERT);
+
+        adv.floodAdvert().txHashes = std::move(mOutgoingTxHashes);
+        auto msg = std::make_shared<StellarMessage>(adv);
+        mOutgoingTxHashes.clear();
+        mApp.postOnMainThread(
+            [weak = mWeakPeer, msg = std::move(msg)]() {
+                auto strong = weak.lock();
+                if (strong)
+                {
+                    strong->sendMessage(msg);
+                }
+            },
+            "flushAdvert");
+    }
+}
+
+void
+TxPullMode::shutdown()
+{
+    mAdvertTimer.cancel();
+}
+
+void
+TxPullMode::startAdvertTimer()
+{
+    mAdvertTimer.expires_from_now(mApp.getConfig().FLOOD_ADVERT_PERIOD_MS);
+    mAdvertTimer.async_wait([this](asio::error_code const& error) {
+        if (!error)
+        {
+            flushAdvert();
+        }
+    });
+}
+
+void
+TxPullMode::queueOutgoingAdvert(Hash const& txHash)
+{
+    if (mOutgoingTxHashes.empty())
+    {
+        startAdvertTimer();
+    }
+
+    mOutgoingTxHashes.emplace_back(txHash);
+
+    // Flush adverts at the earliest of the following two conditions:
+    // 1. The number of hashes reaches the threshold.
+    // 2. The oldest tx hash hash been in the queue for FLOOD_TX_PERIOD_MS.
+    if (mOutgoingTxHashes.size() == getMaxAdvertSize())
```

Comment:
It's a bit misleading since there are two conditions for flushing adverts, one is implemented in this function, and the other is managed elsewhere by the advert time. I edited to comment to clarify this.